### PR TITLE
Add location info to `Rust::Identifier`

### PR DESCRIPTION
**Pull request by tamaroning**
Fixes https://github.com/Rust-GCC/gccrs/issues/2333

```
gcc/rust/ChangeLog:

	* ast/rust-ast.h: Modify constructors of `Rust::Identifier`
	* ast/rust-pattern.h: Likewise.
	* hir/tree/rust-hir-item.h: Likewise.
	* parse/rust-parse-impl.h (Parser::parse_macro_rules_def): Likewise.
	(Parser::parse_decl_macro_def): Likewise.
	(Parser::parse_macro_match_fragment): Likewise.
	(Parser::parse_module): Likewise.
	(Parser::parse_use_tree): Likewise.
	(Parser::parse_function): Likewise.
	(Parser::parse_type_param): Likewise.
	(Parser::parse_type_alias): Likewise.
	(Parser::parse_struct): Likewise.
	(Parser::parse_struct_field): Likewise.
	(Parser::parse_enum): Likewise.
	(Parser::parse_enum_item): Likewise.
	(Parser::parse_union): Likewise.
	(Parser::parse_static_item): Likewise.
	(Parser::parse_trait): Likewise.
	(Parser::parse_trait_item): Likewise.
	(Parser::parse_trait_type): Likewise.
	(Parser::parse_trait_const): Likewise.
	(Parser::parse_external_item): Likewise.
	(Parser::parse_generic_args_binding): Likewise.
	(Parser::parse_method): Likewise.
	(Parser::parse_maybe_named_param): Likewise.
	(Parser::parse_identifier_pattern): Likewise.
	(Parser::parse_struct_expr_field): Likewise.
	(ResolveItem::visit): Likewise.
	* typecheck/rust-hir-type-check-implitem.cc (TypeCheckImplItem::visit):Add comments
	* resolve/rust-ast-resolve-item.cc (ResolveTraitItems::visit): Likewise.
	* checks/lints/rust-lint-scan-deadcode.h: Fix error location
```

### CODE REVIEW

**Comment by tamaroning** [comment]
Now we get errors like the following:
```
$ cat a.rs
fn foo() {}

struct _S ();

impl _S {
  fn bar() {}
}
$ gccrs -frust-incomplete-and-experimental-compiler-do-not-use a.rs
a.rs:1:4: warning: function is never used: ‘foo’
    1 | fn foo() {}
      |    ^~~
a.rs:3:1: warning: struct is never constructed: ‘_S’
    3 | struct _S ();
      | ^~~~~~
a.rs:6:6: warning: associated function is never used: ‘bar’
    6 |   fn bar() {}
      |      ^~~
/usr/bin/ld: /lib/x86_64-linux-gnu/crt1.o: in function `_start':
(.text+0x1b): undefined reference to `main'
collect2: error: ld returned 1 exit status
$
```

File: gcc/rust/checks/lints/rust-lint-scan-deadcode.h [hunk 1]
```diff
@@ -60,15 +60,16 @@ class ScanDeadcode : public MarkLiveBase
...
 		  function.get_function_name ().as_string ().c_str ());
 	      }
 	  }
 	else
 	  {
 	    rust_warning_at (
-	      function.get_locus (), 0, "function is never used: %<%s%>",
+	      function.get_function_name ().get_locus (), 0,
+	      "function is never used: %<%s%>",
 	      function.get_function_name ().as_string ().c_str ());
```
Comment (tamaroning) [inline gcc/rust/checks/lints/rust-lint-scan-deadcode.h:73]: I fixed error location only for check for unused functions.
Of course, other error location should be fixed in another PR.

Comment (philberty) [reply gcc/rust/checks/lints/rust-lint-scan-deadcode.h:73]: Thanks for doing that but you are correct the location info like this is a much bigger issue and outside of scope for your google summer of code project so this PR is more than enough.

**Review by philberty — APPROVED** [review]
LGTM nice work

**Review by P-E-P — APPROVED** [review]

**Comment by tamaroning** [comment]
Thank you for your review :) @P-E-P @philberty
